### PR TITLE
contrib ctc interface changes, cudnn7 CTC, and gluon CTC

### PR DESCRIPTION
**Pull request by szha**
This change is to make current contrib CTC compatible with the cudnn7 CTC interface, and to add CTC loss layer for gluon.

### CODE REVIEW

**Comment by szha** [comment]
@sbodenstein the second commit in this PR is to address #7445.

File: src/operator/contrib/ctc_loss-inl.h [hunk 1]
```diff
@@ -128,40 +133,104 @@ inline void get_workspace_size(std::vector<int> *label_lengths,
...
+#endif
+    std::copy(cpu_labels.begin(), cpu_labels.begin() + len,
+              std::back_inserter(*packed_labels));
+    label_lengths->at(b) = len;
+  }
+  return exceed_limit;
+}
+
+template <typename DType, typename xpu>
+inline bool PackLabelByLength(mshadow::Tensor<xpu, 2, DType> labels,
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss-inl.h]: add comment

File: src/operator/contrib/ctc_loss-inl.h [hunk 2]
```diff
@@ -240,15 +461,22 @@ class CTCLossProp : public OperatorProperty {
   int NumOutputs() const override { return 2; }
 
   std::vector<std::string> ListArguments() const override {
-    return {"data", "label"};
+    if (param_.use_input_lengths && param_.use_label_lengths) {
+      return {"data", "label", "input_lengths", "label_lengths"};
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss-inl.h]: data_lengths

File: tests/python/unittest/test_loss.py [hunk 3]
```diff
@@ -165,6 +165,36 @@ def test_l1_loss():
     assert mod.score(data_iter, eval_metric=mx.metric.Loss())[0][1] < 0.1
 
 
+def test_ctc_loss():
+    loss = gluon.loss.CTCLoss(padding_mask=0)
```
Comment (piiswrong) [inline tests/python/unittest/test_loss.py]: test gpu

File: src/operator/contrib/ctc_loss-inl.h [hunk 4]
```diff
@@ -221,12 +315,139 @@ class CTCLossOp : public Operator {
...
+                                    std::vector<int>* label_lengths,
+                                    std::vector<int>* packed_labels,
+                                    int max_seq_len,
+                                    int batch_size,
+                                    int alphabet_size) {
+    using namespace mshadow;
+    // since the input is activation before softmax and cudnn ctc takes softmax
+    // apply softmax to inputs first.
+    Tensor<xpu, 3, real_t> prob(data.shape_);
+    mshadow::AllocSpace(&prob);
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss-inl.h]: Dont alloc this way. Use ctx.requested

File: tests/python/unittest/test_loss.py (hunk 3 above)
Comment (szha) [inline tests/python/unittest/test_loss.py]: I included test_loss in test_operator_gpu and everything passes.

File: src/operator/contrib/ctc_loss-inl.h [hunk 5]
```diff
@@ -178,27 +252,47 @@ class CTCLossOp : public Operator {
...
+
+    if (param_.use_label_lengths) {
+      int kLabelLength = 2+param_.use_input_lengths;
+#if defined(__CUDACC__) && MXNET_USE_CUDNN == 1 && CUDNN_MAJOR >= 7
+      exceed_cudnn_limit =
+#endif
+      PackLabelByLength(labels, in_data[kLabelLength].get<xpu, 1, real_t>(s),
+                        &packed_labels, &label_lengths);
+    } else {
+#if defined(__CUDACC__) && MXNET_USE_CUDNN == 1 && CUDNN_MAJOR >= 7
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss-inl.h]: remove these ifs

File: src/operator/contrib/ctc_loss-inl.h [hunk 6]
```diff
@@ -125,43 +130,112 @@ inline void get_workspace_size(std::vector<int> *label_lengths,
...
+    CUDNN_CALL(cudnnSetCTCLossDescriptor(ctc_desc_, CUDNN_DATA_FLOAT));
+    CUDNN_CALL(cudnnCreateTensorDescriptor(&prob_desc_));
+    CUDNN_CALL(cudnnCreateTensorDescriptor(&grad_desc_));
+  }
+
+  ~CTCLossOp() {
+    CUDNN_CALL(cudnnDestroyCTCLossDescriptor(ctc_desc_));
+    CUDNN_CALL(cudnnDestroyTensorDescriptor(prob_desc_));
+    CUDNN_CALL(cudnnDestroyTensorDescriptor(grad_desc_));
+#endif
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss-inl.h]: if/endif shouldn't cross function

File: python/mxnet/gluon/loss.py [hunk 7]
```diff
@@ -295,3 +297,102 @@ def hybrid_forward(self, F, output, label, sample_weight=None):
...
+      [[2, 1, 0, 0], [3, 2, 2, 0]]
+
+
+    Parameters
+    ----------
+    output_layout : str, default 'NTC'
+        Layout of the output sequence activation vector.
+    label_layout : str, default 'NT'
+        Layout of the labels.
+    use_input_lengths : bool, default False
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: no need for the flag. just see if argument is none

File: python/mxnet/gluon/loss.py [hunk 8]
```diff
@@ -295,3 +297,102 @@ def hybrid_forward(self, F, output, label, sample_weight=None):
...
+    length. For example, if *label_sequence_length* = 4, and one has two sequences
+    of labels [2, 1] and [3, 2, 2], the resulting ```label``` tensor should be
+    padded to be::
+
+      [[2, 1, 0, 0], [3, 2, 2, 0]]
+
+
+    Parameters
+    ----------
+    output_layout : str, default 'NTC'
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: layout

File: python/mxnet/gluon/loss.py [hunk 9]
```diff
@@ -295,3 +297,102 @@ def hybrid_forward(self, F, output, label, sample_weight=None):
...
+    use_label_lengths : bool, default False
+        Whether to use `label_lengths` to decide lengths of labels.
+        If false, the label lengths are derived from the first occurrence of
+        the value specified by `padding_mask`.
+    padding_mask : int or None, default -1
+        This is the label value to be considered padding, which is used to derive the actual
+        lengths of labels. Only required when `use_label_lengths` is false.
+    weight : float or None
+        Global scalar weight for loss.
+    input_lengths : NDArray or None,
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: these are argument to forward. List separately in input/output section

File: src/operator/contrib/ctc_loss-inl.h [hunk 10]
```diff
@@ -125,43 +130,112 @@ inline void get_workspace_size(std::vector<int> *label_lengths,
...
+// When cudnn is enabled, the return value signifies whether the cudnn length limit is exceeded.
+template <typename DType, typename xpu>
+inline bool PackLabelByLength(mshadow::Tensor<xpu, 2, DType> labels,
+                              mshadow::Tensor<xpu, 1, DType> in_label_lengths,
+                              std::vector<int> *packed_labels,
+                              std::vector<int> *label_lengths) {
+  int batch = labels.size(0);
+  int max_num_labels = labels.size(1);
+  std::vector<int> cpu_labels(max_num_labels);
+  IndexTensorToVector(in_label_lengths, label_lengths);
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss-inl.h]: this function does cudaMemcopy.
Can you do it only once?

File: src/operator/contrib/ctc_loss-inl.h [hunk 11]
```diff
@@ -125,43 +130,112 @@ inline void get_workspace_size(std::vector<int> *label_lengths,
...
                                       std::vector<int> *packed_labels,
                                       std::vector<int> *label_lengths) {
   int batch = labels.size(0);
   int max_num_labels = labels.size(1);
-  std::vector<index_t> cpu_labels(max_num_labels);
+  std::vector<int> cpu_labels(max_num_labels);
+  bool exceed_limit = false;
 
   for (int b = 0; b < batch; ++b) {
     IndexTensorToVector(labels[b], &cpu_labels);
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss-inl.h]: try to do copy only once

**Comment by piiswrong** [comment]
Do you have performance comp between baidu and cudnn?

**Comment by szha** [comment]
I compared these two implementations roughly on K80 by timing only the parts where the two implementation diverges. The two implementations are about the same, and cudnn one seems slightly more efficient when input size is large. Below I listed the numbers. The four parts of the workload line represent: input shape before preprocessing, label shape, input length, label length.
```
workload: (32L, 256L, 28L), (32L, 1L), 256, 1
baidu
forward elapsed time: 614598 ns
cudnn
forward elapsed time: 545639 ns

workload: (32L, 512L, 28L), (32L, 1L), 512, 1
baidu
forward elapsed time: 1030004 ns
cudnn
forward elapsed time: 1063012 ns

workload: (32L, 1024L, 28L), (32L, 1L), 1024, 1
baidu
forward elapsed time: 2115799 ns
cudnn
forward elapsed time: 2207311 ns

workload: (32L, 2048L, 28L), (32L, 1L), 2048, 1
baidu
forward elapsed time: 4085191 ns
cudnn
forward elapsed time: 4085597 ns

workload: (32L, 256L, 28L), (32L, 2L), 256, 2
baidu
forward elapsed time: 680476 ns
cudnn
forward elapsed time: 685418 ns

workload: (32L, 512L, 28L), (32L, 2L), 512, 2
baidu
forward elapsed time: 1387201 ns
cudnn
forward elapsed time: 1338115 ns

workload: (32L, 1024L, 28L), (32L, 2L), 1024, 2
baidu
forward elapsed time: 11913695 ns
cudnn
forward elapsed time: 4217115 ns

workload: (32L, 2048L, 28L), (32L, 2L), 2048, 2
baidu
forward elapsed time: 7950573 ns
cudnn
forward elapsed time: 5184110 ns

workload: (32L, 256L, 28L), (32L, 4L), 256, 4
baidu
forward elapsed time: 950801 ns
cudnn
forward elapsed time: 952418 ns

workload: (32L, 512L, 28L), (32L, 4L), 512, 4
baidu
forward elapsed time: 1940807 ns
cudnn
forward elapsed time: 1873934 ns

workload: (32L, 1024L, 28L), (32L, 4L), 1024, 4
baidu
forward elapsed time: 3801619 ns
cudnn
forward elapsed time: 3755733 ns

workload: (32L, 2048L, 28L), (32L, 4L), 2048, 4
baidu
forward elapsed time: 7530379 ns
cudnn
forward elapsed time: 7496358 ns

workload: (32L, 256L, 28L), (32L, 8L), 256, 8
baidu
forward elapsed time: 1474249 ns
cudnn
forward elapsed time: 1472007 ns

workload: (32L, 512L, 28L), (32L, 8L), 512, 8
baidu
forward elapsed time: 2975453 ns
cudnn
forward elapsed time: 2951099 ns

workload: (32L, 1024L, 28L), (32L, 8L), 1024, 8
baidu
forward elapsed time: 7873482 ns
cudnn
forward elapsed time: 7956207 ns

workload: (32L, 2048L, 28L), (32L, 8L), 2048, 8
baidu
forward elapsed time: 14353415 ns
cudnn
forward elapsed time: 11856699 ns

workload: (32L, 256L, 28L), (32L, 16L), 256, 16
baidu
forward elapsed time: 2529732 ns
cudnn
forward elapsed time: 2527600 ns

workload: (32L, 512L, 28L), (32L, 16L), 512, 16
baidu
forward elapsed time: 5129444 ns
cudnn
forward elapsed time: 5202983 ns

workload: (32L, 1024L, 28L), (32L, 16L), 1024, 16
baidu
forward elapsed time: 20691226 ns
cudnn
forward elapsed time: 10203190 ns

workload: (32L, 2048L, 28L), (32L, 16L), 2048, 16
baidu
forward elapsed time: 20736001 ns
cudnn
forward elapsed time: 20764510 ns

workload: (32L, 256L, 28L), (32L, 32L), 256, 32
baidu
forward elapsed time: 4349291 ns
cudnn
forward elapsed time: 4372212 ns

workload: (32L, 512L, 28L), (32L, 32L), 512, 32
baidu
forward elapsed time: 9181241 ns
cudnn
forward elapsed time: 9177710 ns

workload: (32L, 1024L, 28L), (32L, 32L), 1024, 32
baidu
forward elapsed time: 26316193 ns
cudnn
forward elapsed time: 19705820 ns

workload: (32L, 2048L, 28L), (32L, 32L), 2048, 32
baidu
forward elapsed time: 37893973 ns
cudnn
forward elapsed time: 37927728 ns

workload: (32L, 256L, 28L), (32L, 64L), 256, 64
baidu
forward elapsed time: 7350089 ns
cudnn
forward elapsed time: 7341236 ns

workload: (32L, 512L, 28L), (32L, 64L), 512, 64
baidu
forward elapsed time: 10367995 ns
cudnn
forward elapsed time: 10394812 ns

workload: (32L, 1024L, 28L), (32L, 64L), 1024, 64
baidu
forward elapsed time: 26141244 ns
cudnn
forward elapsed time: 26080623 ns

workload: (32L, 2048L, 28L), (32L, 64L), 2048, 64
baidu
forward elapsed time: 46190315 ns
cudnn
forward elapsed time: 43888343 ns

workload: (32L, 256L, 28L), (32L, 128L), 256, 128
baidu
forward elapsed time: 7244567 ns
cudnn
forward elapsed time: 7197396 ns

workload: (32L, 512L, 28L), (32L, 128L), 512, 128
baidu
forward elapsed time: 17969224 ns
cudnn
forward elapsed time: 17975323 ns

workload: (32L, 1024L, 28L), (32L, 128L), 1024, 128
baidu
forward elapsed time: 37417761 ns
cudnn
forward elapsed time: 37404967 ns

workload: (32L, 2048L, 28L), (32L, 128L), 2048, 128
baidu
forward elapsed time: 134485276 ns
cudnn
forward elapsed time: 134316760 ns

workload: (32L, 256L, 28L), (32L, 256L), 256, 256
baidu
forward elapsed time: 7399992 ns
cudnn
forward elapsed time: 7459652 ns

workload: (32L, 512L, 28L), (32L, 256L), 512, 256
baidu
forward elapsed time: 20783582 ns
cudnn
forward elapsed time: 20854272 ns

workload: (32L, 1024L, 28L), (32L, 256L), 1024, 256
baidu
forward elapsed time: 67250512 ns
cudnn
forward elapsed time: 59459103 ns

workload: (32L, 2048L, 28L), (32L, 256L), 2048, 256
baidu
forward elapsed time: 147523944 ns
cudnn
forward elapsed time: 145160144 ns
```
